docs/rfc/rfc-2.adoc

```diff
@@ -0,0 +1,799 @@
+
+To avoid _strictly requiring_ a connection to a Bitcoin node, this RFC
+proposes tracking the latest address generated, and allowing an additional
+config key, `TargetDerivationStartIndex`, that sets a starting point for such
+derivations. The client should write the updated index to the `current/`
+directory that tracks active key shares, to an encrypted file named
+`current-btc-refund-derivation-index`. The operator should then seek to
+obtain from the staker an extended public key that is only used for this
+purpose. In cases where in-memory/on-disk derivation index and config file index
+both exist, the highest index should be used.
+
+If an optional Bitcoin node connection is available (see <<electrs>>), the
+client should additionally check a derived address for a balance before
+publishing. To do this, a GET request to `/address/<address>` should be issued,
+and the JSON response inspected for a `tx_count` of 0. When such a connection is
+available, if the client finds an address is already used, it should increment
+the derivation index and check again. If the GET request fails, the index should
+be assumed to be valid, as if the Bitcoin node connection were not available.
+
+==== Modification of key material archiving
+
+In the current client, key material is archived as soon as the ECDSA keep
+contract a client is participating in moves into the CLOSED or TERMINATED
+state. This RFC revises that behavior to only archive on transition to CLOSED.
+When contracts transition into TERMINATED state and that transition is
+confirmed (by waiting
+https://github.com/keep-network/keep-ecdsa/blob/master/pkg/client/client.go#L28-L31[`blockConfirmation`
+blocks]), the client should immediately begin <<bitcoin-fee,Bitcoin fee
+resolution>> followed by the <<signing-protocol>> and finally
+<<transaction-submission,transaction submission>>. Only when signing and
+transaction submission (whether via logging or API) completes or times out
+should the associated key material be archived.
+
+On startup, if a client sees that an unarchived keep is in the TERMINATED
+state and has been in that state for `blockConfirmation` blocks, it should
+move directly to <<bitcoin-fee,Bitcoin fee resolution>>, unless an optional
+Bitcoin node connection is available (see <<electrs>>). In this case, the
+client should check the keep's associated deposit for an unspent balance by
+issuing a GET request to `/tx/:txid/outspend/:vout`, using the deposit's
+`fundingInfo().utxoOutpoint` to derive the transaction id and vout, and only
+start the signing process if the JSON result has a `spent` value of false. If
+`spent` is `true` and `status.confirmed` is `false`, a retry should occur
+after an hour until `spent` is `true`, `status.confirmed` is `true`, and
+`status.block_height` is at least 6 blocks past the value returned by issuing
+a GET request to `/blocks/tip/height`. If this occurs, the key material
+should be archived.
```

Comment:
There is a scenario possible when one node opts out from recovery and two other report errors because they do not have electrs configured. To address it, we could be logging a warning that electrs is not configured and the client attempts signing without executing the check.

Reply:
Not sure I understand this scenario all the way. If one node opts out from recovery altogether (meaning it either has an older version of the client, hasn't configured a return address, or is on a custom build that has omitted the recovery functionality), that will lead to errors when the signing protocol starts, rather than here.

If electrs isn't configured (or, in the current RFC's state, if it is _explicitly disabled_), then we won't try to use it. It could be misconfigured, of course. Either way though, we should log errors if there are errors interacting with electrs.

Reply:
Right, that's what I meant. Say we have three clients, two of them do not have electrs configured and hence do not perform the check. One client has electrs configured, it performs the check and discovers BTC has been spent and there is no point in joining the signing process.

In this scenario, we have two clients attempting to sign the transaction and reporting errors that the third client is not responding. 

My suggestion was to print a warning to logs on those two clients before they enter the signing process saying  that electrs is not configured, they can not perform the preliminary check of the deposit, and they enter the process regardless. This could help explain why the process is failing.

Reply:
WARNing that electrs isn't configured makes sense, but I wonder if we should just do it once at startup? Right now we [recommend](https://docs.keep.network/monitoring-and-alerting.html) that more than 10 WARN logs in 15 minutes should trigger an alert. If there's a situation where 10 liquidations trigger around the same time due to price movement, we would trigger an alert, which doesn't feel like the correct thing in this scenario.

Reply:
Yeah, emitting them once at startup makes sense. We just need to make sure they clearly articulate what are the implications.